### PR TITLE
Universal Train Data Generator

### PR DESCRIPTION
**Pull request by yoonichoi**
uses ```setup_dict``` to create a train data using GPT-J 6B.

```setup_dict``` should include: 
- train_path (str): path to the train data
- template_dict (dict): tp_dict to be passed into `PromptTemplate`
- data_columns (dict): input columns and output column to be used in `DatasetReader`
- label_map (dict): mapping of label (key) to the label string (value)
- output_path (str): path to store the created train data

~~Still in progress to automatically incorporate:~~
~~- ```p['label_text']```: probably should also provided from ```setup_dict```~~
~~- conversion of PPL > probabilities~~

**3/13 update**
- supports label_text mapping
- added `pseudo_gt` key to the generated train data as well (for further usage during KD stage)
- supports PPL -> probability conversion (*may want to switchover to prob_inferencer later*)
command line usage:
`make run-create_train SETUP_DICT=<path/to/setupdict_json>`

### CODE REVIEW

File: scripts/create_train.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                yield json.loads(line)
+    
+    train_ds = Dataset.from_generator(gen, gen_kwargs={"file_path": train_path})
+    val_ds = Dataset.from_generator(gen, gen_kwargs={"file_path": train_path})
+    test_ds = Dataset.from_generator(gen, gen_kwargs={"file_path": train_path})
+    
+    dataset_dict = DatasetDict({"train": train_ds, "validation": val_ds, "test": test_ds})
+    dataset = dataset_dict
+    
+    data = DatasetReader(dataset, input_columns=['text'], output_column='label')
```
Comment (DHdroid) [inline scripts/create_train.py]: json 파일로 뺴는거 반영 부탁드립니다!

File: scripts/create_train.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import numpy as np
+
+def rec_softmax(x):
+    e_x = np.exp(x)
+    e_x = 1/e_x
+    return e_x / e_x.sum(axis=0)
+
+
+vessl.init()
```
Comment (DHdroid) [inline scripts/create_train.py]: 삭제부탁드립니다!

File: scripts/create_train.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    
+    
+    template = PromptTemplate(tp_dict, {'text': '</text>'}, ice_token='</E>')
+    retriever = RandomRetriever(data, ice_num=0)
+    inferencer = ParentInferencer(model_name='EleutherAI/gpt-j-6b')
+    predictions = inferencer.inference(retriever, ice_template=template)
+    
+    for i, p in enumerate(predictions):
+        entry = dataset_dict["test"][i]
+        p["text"] = entry["text"]
```
Comment (DHdroid) [inline scripts/create_train.py]: 이거 input_columns하고 동일한가요? 아니면 항상 text인가요?

File: Makefile [hunk 4]
```diff
@@ -25,6 +25,9 @@ run-create_data_sst2:
 run-create_subj:
 	$(AUTO_POETRY) python scripts/create_subj_train.py
 
+run-create_train:
```
Comment (DHdroid) [inline Makefile:17]: 실제 make로 실행할 때는 setup dict 어떻게 넘겨주나요? 환경변수 사용해야하나요?

File: .DS_Store
Comment (nemodleo) [inline .DS_Store:1]: 제거

File: data/.DS_Store
Comment (nemodleo) [inline data/.DS_Store:1]: 동일

File: Makefile [hunk 5]
```diff
@@ -19,6 +19,15 @@ run-sst5:
 run-ag_news:
 	$(AUTO_POETRY) python scripts/ag_news.py
 
+run-create_data_sst2:
+	$(AUTO_POETRY) python scripts/create_traindata_sst2.py
+
+run-create_subj:
```
Comment (nemodleo) [inline Makefile]: create_data_sst2와 통일성을 맞추는게 좋아보입니다

File: Makefile [hunk 6]
```diff
@@ -19,6 +19,15 @@ run-sst5:
...
+	$(AUTO_POETRY) python scripts/create_subj_train.py
```
Comment (nemodleo) [inline Makefile]: 동일

File: Makefile (hunk 4 above)
Comment (nemodleo) [reply Makefile:17]: example
```
make run-create_train \
    args_a=blabla \
    args_b=blablabla ...
```
&
```
run-create_train:
    $(AUTO_POETRY) python scripts/create_train.py --ARGS_A=$(args_a) --ARGS_B=$(args_b)
```

File: data/sst2/.DS_Store
Comment (nemodleo) [inline data/sst2/.DS_Store:1]: 동일

File: data_utils/.DS_Store
Comment (nemodleo) [inline data_utils/.DS_Store:1]: 동일

File: data_utils/create_traindata_sst5.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+from datasets import load_dataset
+from datasets import Dataset, DatasetDict
+from openicl import DatasetReader
+import json
+import vessl
+
+# Before : preprocess data !
+
+vessl.init()
```
Comment (nemodleo) [inline data_utils/create_traindata_sst5.py]: 삭제하기로 했습니다

File: data_utils/create_traindata_sst5.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+template = PromptTemplate(tp_dict, {'text': '</text>'}, ice_token='</E>')
+
+
+from openicl import RandomRetriever
+# Define a retriever using the previous `DataLoader`.
+# `ice_num` stands for the number of data in in-context examples.
+retriever = RandomRetriever(data, ice_num=0)
+
+import ParentInferencer
+inferencer = ParentInferencer.ParentInferencer(model_name='EleutherAI/gpt-j-6b')
```
Comment (nemodleo) [inline data_utils/create_traindata_sst5.py]: 6b -> 6B,

File: data_utils/create_traindata_sst5.py
Comment (nemodleo) [inline data_utils/create_traindata_sst5.py:1]: 이파일은 develop꺼 사용하면 될듯합니다

File: scripts/sst5.py [hunk 9]
```diff
@@ -1,52 +1,255 @@
...
+    # compute accuracy for the prediction
     score = AccEvaluator().score(predictions=predictions, references=data.references)
-    print(score)
+    
+    return score
 
 
 if __name__ == '__main__':
-    test()
+    test()
```
Comment (nemodleo) [inline scripts/sst5.py]: 되도록 파일 올리기전에 make check-quality 합시다!

File: iclx/inferencer/base_inferencer.py [hunk 10]
```diff
@@ -83,15 +83,7 @@ def inference(self,
...
-        model_config = AutoConfig.from_pretrained(model_name)
-        with init_empty_weights():
-            empty_model = AutoModelForCausalLM.from_config(model_config)
-        device_map = infer_auto_device_map(empty_model, dtype="float16")
-        self.model = AutoModelForCausalLM.from_pretrained(model_name,
-                                                          device_map=device_map,
-                                                          offload_folder="offload",
-                                                          offload_state_dict=True,
-                                                          torch_dtype=torch.float16)
+        self.model = AutoModelForCausalLM.from_pretrained(model_name)
```
Comment (nemodleo) [inline iclx/inferencer/base_inferencer.py]: 해당 브랜치 develop rebase 하고 리뷰하는게 좋아보입니다. 
 --- 이하 리뷰 스테쉬

File: Makefile (hunk 4 above)
Comment (DHdroid) [reply Makefile:17]: 그럼 argparse하나 추가할 때마다 좀 귀찮긴 하네요 😞

Comment (yoonichoi) [reply Makefile:17]: 기존방식대로진행할지, 아니면 네모네모님 의견대로 진행할지 내일 이야기 나눠보시죠!

File: scripts/sst5.py (hunk 9 above)
Comment (DHdroid) [reply scripts/sst5.py]: @nemodleo 요거 근데 기존 파일들(openicl에 있던)도 mypy랑 flake에 다 걸리던데 잘 쓰는 방법 아시나요?
